### PR TITLE
feat(App): add message & notification options (#40456)

### PR DESCRIPTION
**Pull request by luo3house**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

close https://github.com/ant-design/ant-design/issues/40456

### 💡 Background and solution

New options `messageConfig` and `notificationConfig`, forwarded from App wrapper

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |  add configurable message and notification options for App  |
| 🇨🇳 Chinese | App 提供预先配置 message、notification 的选项  |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by MadCcc** [comment]
feature branch plz.

**Comment by luo3house** [comment]
I dont think the check went through the commit, any way to re-run checks? It is now ready.

**Comment by Wxh16144** [comment]
a question.

Can the config set via the `<App />` component also take effect for static methods? This way we can use only one configuration. Lower the priority of the original `message.config`.

**Comment by luo3house** [comment]
> a question.
> 
> Can the config set via the `<App />` component also take effect for static methods? This way we can use only one configuration. Lower the priority of the original `message.config`.

I think it may be cause unexpected global config overwritten in some cases about nested `<App`. 

The `message` provided by `useApp()` is configured by nearest `App` wrapper, as a scoped API.

**Comment by MadCcc** [comment]
How about inherit configruation if App is nested? Consider that we won't want parent configuration broken.

**Comment by luo3house** [comment]
> How about inherit configruation if App is nested? Consider that we won't want parent configuration broken.

lgtm, for this scenario, AppContext holds configurations instead of early created message, noti and modal instance

File: components/app/index.tsx [hunk 1]
```diff
@@ -15,26 +17,59 @@ export type AppProps = {
...
+} & AppConfig;
+
+export type AppConfig = {
+  messageConfig?: MessageConfig;
+  notificationConfig?: NotificationConfig;
+};
+
+export type InternalUseAppProps = useAppProps & {
+  /** @internal: internal usage only for nested app */
+  __INTERNAL__?: AppConfig;
```
Comment (MadCcc) [inline components/app/index.tsx:29]: What's this for?

Comment (luo3house) [reply components/app/index.tsx:29]: The internal Props is used between parent App and child App, to ensure child App can receive options from parent, and merge them, also forward to its nested children Apps.

Comment (MadCcc) [reply components/app/index.tsx:29]: I prefer not to have this prop, merged context is enough.

Comment (nuintun) [reply components/app/index.tsx:29]: @MadCcc 如果定义组件不可见，建议去除 className 属性和 div 层。

Comment (MadCcc) [reply components/app/index.tsx]: > @MadCcc 如果定义 App 组件不可见，建议去除 className 属性和 div 层。

`style` 现在看来还是有必要，在另一个 PR 里处理吧。

File: components/app/index.en-US.md [hunk 2]
```diff
@@ -117,3 +117,12 @@ export default () => {
...
 ```
+
+## API
+
+### App
+
+| Property | Description | Type | Default | Version |
+| --- | --- | --- | --- | --- |
+| messageConfig | Global config for Message | [MessageConfig](/components/message/#messageconfig) | - |  |
+| notificationConfig | Global config for Notification | [NotificationConfig](/components/notification/#notificationconfig) | - |  |
```
Comment (MadCcc) [inline components/app/index.en-US.md:128]:
```suggestion
| message | Global config for Message | [MessageConfig](/components/message/#messageconfig) | - |  |
| notification | Global config for Notification | [NotificationConfig](/components/notification/#notificationconfig) | - |  |
```

File: components/app/index.zh-CN.md [hunk 3]
```diff
@@ -118,3 +118,12 @@ export default () => {
...
 ```
+
+## API
+
+### App
+
+| 参数 | 说明 | 类型 | 默认值 | 版本 |
+| --- | --- | --- | --- | --- |
+| messageConfig | App 内 Message 的全局配置 | [MessageConfig](/components/message-cn/#messageconfig) | - |  |
+| notificationConfig | App 内 Notification 的全局配置 | [NotificationConfig](/components/notification-cn/#notificationconfig) | - |  |
```
Comment (MadCcc) [inline components/app/index.zh-CN.md:129]:
```suggestion
| message | App 内 Message 的全局配置 | [MessageConfig](/components/message-cn/#messageconfig) | - |  |
| notification | App 内 Notification 的全局配置 | [NotificationConfig](/components/notification-cn/#notificationconfig) | - |  |
```

**Comment by MadCcc** [comment]
LGTM, please rebase feature

File: components/app/index.tsx [hunk 4]
```diff
@@ -6,28 +6,46 @@ import { ConfigContext } from '../config-provider';
...
   const [wrapSSR, hashId] = useStyle(prefixCls);
   const customClassName = classNames(hashId, prefixCls, className, rootClassName);
 
-  const [messageApi, messageContextHolder] = useMessage();
-  const [notificationApi, notificationContextHolder] = useNotification();
+  const appConfig = useContext(AppConfigContext);
+  const mergedAppConfig = React.useMemo<AppConfig>(
+    () => ({
+      message: { ...message, ...appConfig.message },
+      notification: { ...notification, ...appConfig.notification },
```
Comment (MadCcc) [inline components/app/index.tsx:40]:
```suggestion
      message: { ...appConfig.message, ...message  },
      notification: { ...appConfig.notification, ...notification, },
```
Current props should be respected. Add a test case for it?

File: components/app/index.zh-CN.md [hunk 5]
```diff
@@ -119,3 +119,12 @@ export default () => {
...
 ```
+
+## API
+
+### App
+
+| 参数 | 说明 | 类型 | 默认值 | 版本 |
+| --- | --- | --- | --- | --- |
+| message | App 内 Message 的全局配置 | [MessageConfig](/components/message-cn/#messageconfig) | - |  |
+| notification | App 内 Notification 的全局配置 | [NotificationConfig](/components/notification-cn/#notificationconfig) | - |  |
```
Comment (MadCcc) [inline components/app/index.zh-CN.md:130]:
```suggestion
| message | App 内 Message 的全局配置 | [MessageConfig](/components/message-cn/#messageconfig) | - | 5.3.0 |
| notification | App 内 Notification 的全局配置 | [NotificationConfig](/components/notification-cn/#notificationconfig) | - | 5.3.0 |
```

File: components/app/index.en-US.md [hunk 6]
```diff
@@ -118,3 +118,12 @@ export default () => {
...
 ```
+
+## API
+
+### App
+
+| Property | Description | Type | Default | Version |
+| --- | --- | --- | --- | --- |
+| message | Global config for Message | [MessageConfig](/components/message/#messageconfig) | - |  |
+| notification | Global config for Notification | [NotificationConfig](/components/notification/#notificationconfig) | - |  |
```
Comment (MadCcc) [inline components/app/index.en-US.md:129]:
```suggestion
| message | Global config for Message | [MessageConfig](/components/message/#messageconfig) | - | 5.3.0 |
| notification | Global config for Notification | [NotificationConfig](/components/notification/#notificationconfig) | - | 5.3.0 |
```

**Review by MadCcc — APPROVED** [review]

**Pull request merged**